### PR TITLE
dataframe.astype() exception message to include column name

### LINKED ISSUE #47571: ENH: dataframe.astype() exception message to include column name
#### Is your feature request related to a problem?

When using dataframe.astype() and there is a column missing or there are some values that cannot be correctly converted, there is an exception raised. But there is no way to know which column is missing or which column has the wrong value.

#### Describe the solution you'd like

The exception message of the dataframe.astype() should contain the column name responsible for the exception (missing column(s) or column with wrong values).

#### API breaking implications

This does not introduce any breaking changes

#### Additional context

Some examples of the current exceptions raised by astype():

```
KeyError: 'Only a column name can be used for the key in a dtype mappings argument.'

TypeError: object cannot be converted to an IntegerDtype

ValueError: invalid literal for int() with base 10: 'test'
```

### PR DESCRIPTION
**Pull request by dannyi96**
- [X] closes #47571(Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

**Description:**
Added column related info as part of `astype` exceptions

| Code  | Old Error | New Error |
| ------------- | ------------- | ----------|
| pd.DataFrame({'a':[1,2,"text"], 'b':[1,2,3]}).astype({'a': int}) | ValueError: invalid literal for int() with base 10: 'text' | ValueError: invalid literal for int() with base 10: 'text': Error while type casting for column 'a' |

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -6224,12 +6224,18 @@ def astype(
 
             results = []
             for i, (col_name, col) in enumerate(self.items()):
-                cdt = dtype_ser.iat[i]
-                if isna(cdt):
-                    res_col = col.copy() if copy else col
-                else:
-                    res_col = col.astype(dtype=cdt, copy=copy, errors=errors)
-                results.append(res_col)
+                try:
```
Comment (mroeschke) [inline pandas/core/generic.py:6227]: 1. Better if this `try/except` only applies to the line where we know the conversion fails
2. Could the more specific `Exception`s be captured?

**Review by phofl — COMMENTED** [review]
Can you address review comments? Also we need a whatsnew

**Comment by dannyi96** [comment]
thanks for the review @phofl @mroeschke 
sure @phofl, will update the PR with changes

File: pandas/core/generic.py (hunk 1 above)
Comment (dannyi96) [reply pandas/core/generic.py]: done

File: doc/source/whatsnew/v1.6.0.rst [hunk 2]
```diff
@@ -35,7 +35,7 @@ Other enhancements
 - Added ``index`` parameter to :meth:`DataFrame.to_dict` (:issue:`46398`)
 - Added metadata propagation for binary operators on :class:`DataFrame` (:issue:`28283`)
 - :class:`.CategoricalConversionWarning`, :class:`.InvalidComparison`, :class:`.InvalidVersion`, :class:`.LossySetitemError`, and :class:`.NoBufferPresent` are now exposed in ``pandas.errors`` (:issue:`27656`)
--
+- :func:`DataFrame.astype()` exception message thrown improved to include column name when type conversion is not possible. (:issue:`47571`)
```
Comment (dannyi96) [inline doc/source/whatsnew/v1.6.0.rst:38]: not sure if this section is best location in whatsnew for this change

Comment (mroeschke) [inline doc/source/whatsnew/v1.6.0.rst:38]:
```suggestion
- :func:`DataFrame.astype` exception message thrown improved to include column name when type conversion is not possible. (:issue:`47571`)
```

**Comment by dannyi96** [comment]
Hi @mroeschke , had missed following up on this PR.
feel free to let me know if any changes are required here (or) if we should hold this PR for the time being.

Thanks in advance :)

File: pandas/core/generic.py [hunk 3]
```diff
@@ -6261,7 +6261,13 @@ def astype(
                 if isna(cdt):
                     res_col = col.copy() if copy else col
                 else:
-                    res_col = col.astype(dtype=cdt, copy=copy, errors=errors)
+                    try:
+                        res_col = col.astype(dtype=cdt, copy=copy, errors=errors)
+                    except ValueError as ex:
+                        ex.args = (
+                            f"Error during type conversion for column {col_name}: {ex}",
```
Comment (mroeschke) [inline pandas/core/generic.py:6268]: Actually could the `Error during type conversion for column {col_name}:` after the original exception message. Might help users that are trying to catch the original message and it expect it to be first.

Comment (dannyi96) [reply pandas/core/generic.py:6268]: sure, that makes sense. Will update the PR

Comment (dannyi96) [reply pandas/core/generic.py]: done

**Review by phofl — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @dannyi96

**Pull request merged**